redis/src/aio/connection.rs

```diff
@@ -256,109 +258,350 @@ where
+
+        self.resp_rx
+            .next()
             .await
+            .ok_or(next_response_err())
+            .and_then(|value| from_redis_value(&value))
     }
 
-    /// Returns [`Stream`] of [`Msg`]s from this [`PubSub`]s subscriptions.
-    ///
-    /// The message itself is still generic and can be converted into an appropriate type through
-    /// the helper methods on it.
-    pub fn on_message(&mut self) -> impl Stream<Item = Msg> + '_ {
-        ValueCodec::default()
-            .framed(&mut self.0.con)
-            .filter_map(|msg| Box::pin(async move { Msg::from_value(&msg.ok()?.ok()?) }))
+    /// Subscribes to a new channel.
+    pub async fn subscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
+        self.query(cmd("SUBSCRIBE").arg(channel)).await
     }
 
-    /// Returns [`Stream`] of [`Msg`]s from this [`PubSub`]s subscriptions consuming it.
-    ///
-    /// The message itself is still generic and can be converted into an appropriate type through
-    /// the helper methods on it.
-    /// This can be useful in cases where the stream needs to be returned or held by something other
-    /// than the [`PubSub`].
-    pub fn into_on_message(self) -> impl Stream<Item = Msg> {
-        ValueCodec::default()
-            .framed(self.0.con)
-            .filter_map(|msg| Box::pin(async move { Msg::from_value(&msg.ok()?.ok()?) }))
+    /// Subscribes to a new channel with a pattern.
+    pub async fn psubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
+        self.query(cmd("PSUBSCRIBE").arg(pchannel)).await
     }
 
-    /// Exits from `PubSub` mode and converts [`PubSub`] into [`Connection`].
-    pub async fn into_connection(mut self) -> Connection<C> {
-        self.0.exit_pubsub().await.ok();
+    /// Unsubscribes from a channel.
+    pub async fn unsubscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
+        self.query(cmd("UNSUBSCRIBE").arg(channel)).await
+    }
 
-        self.0
+    /// Unsubscribes from a channel with a pattern.
+    pub async fn punsubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
+        self.query(cmd("PUNSUBSCRIBE").arg(pchannel)).await
+    }
+}
+
+pin_project! {
+    /// Represents the receiving side of a pubsub connection.
+    #[project = SplitPubSubStreamProj]
+    pub struct SplitPubSubStream<C = Pin<Box<dyn AsyncStream + Send + Sync>>> {
+        #[pin]
+        stream: SplitStream<Framed<C, ValueCodec>>,
+        #[pin]
+        resp_tx: mpsc::Sender<Value>,
+        resp_waiting: Option<Value>,
+    }
+}
+
+impl<C> SplitPubSubStream<C>
+where
+    C: Unpin + AsyncRead + Send,
+{
+    fn new(stream: SplitStream<Framed<C, ValueCodec>>, resp_tx: mpsc::Sender<Value>) -> Self {
+        SplitPubSubStream {
+            stream,
+            resp_tx,
+            resp_waiting: None,
+        }
+    }
+}
+
+fn map_response_err<T>(value: Result<T, SendError>) -> Result<T, RedisError> {
```

Comment:
the name doesn't show that this function is used only for forwarding errors from stream to sink.